datagateway_api/src/main.py

```diff
@@ -0,0 +1,178 @@
+        )
+        api.add_resource(get_count_endpoint_resource, f"/{entity_name.lower()}/count")
+        spec.path(resource=get_count_endpoint_resource, api=api)
+
+        get_find_one_endpoint_resource = get_find_one_endpoint(
+            entity_name, endpoints[entity_name], backend,
+        )
+        api.add_resource(
+            get_find_one_endpoint_resource, f"/{entity_name.lower()}/findone",
+        )
+        spec.path(resource=get_find_one_endpoint_resource, api=api)
+
+    # Session endpoint
+    session_endpoint_resource = session_endpoints(backend)
+    api.add_resource(session_endpoint_resource, "/sessions")
+    spec.path(resource=session_endpoint_resource, api=api)
+
+    # Table specific endpoints
+    instrument_facility_cycle_resource = instrument_facility_cycles_endpoint(backend)
+    api.add_resource(
+        instrument_facility_cycle_resource, "/instruments/<int:id_>/facilitycycles",
+    )
+    spec.path(resource=instrument_facility_cycle_resource, api=api)
+
+    count_instrument_facility_cycle_res = count_instrument_facility_cycles_endpoint(
+        backend,
+    )
+    api.add_resource(
+        count_instrument_facility_cycle_res,
+        "/instruments/<int:id_>/facilitycycles/count",
+    )
+    spec.path(resource=count_instrument_facility_cycle_res, api=api)
+
+    instrument_investigation_resource = instrument_investigation_endpoint(backend)
+    api.add_resource(
+        instrument_investigation_resource,
+        "/instruments/<int:instrument_id>/facilitycycles/<int:cycle_id>/investigations",
+    )
+    spec.path(resource=instrument_investigation_resource, api=api)
+
+    count_instrument_investigation_resource = count_instrument_investigation_endpoint(
+        backend,
+    )
+    api.add_resource(
+        count_instrument_investigation_resource,
+        "/instruments/<int:instrument_id>/facilitycycles/<int:cycle_id>/investigations"
+        "/count",
+    )
+    spec.path(resource=count_instrument_investigation_resource, api=api)
+
+
+def openapi_config(spec):
+    # Reorder paths (e.g. get, patch, post) so openapi.yaml only changes when there's a
+    # change to the Swagger docs, rather than changing on each startup
+    if config.is_generate_swagger():
+        log.debug("Reordering OpenAPI docs to alphabetical order")
+        for entity_data in spec._paths.values():
+            for endpoint_name in sorted(entity_data.keys()):
+                entity_data.move_to_end(endpoint_name)
+
+    openapi_spec_path = Path(__file__).parent / "swagger/openapi.yaml"
+    with open(openapi_spec_path, "w") as f:
+        f.write(spec.to_yaml())
+
+
+@app.route("/openapi.json")
+def specs():
+    resp = app.make_response(json.dumps(spec.to_dict(), indent=2))
+    resp.mimetype = "application/json"
+    return resp
+
+
+if __name__ == "__main__":
+    api, spec = create_app_infrastructure(app)
+    create_api_endpoints(app, api, spec)
+    openapi_config(spec)
```

Comment:
These lines need to be removed from the `__name__ == "__main__"` section, as when we are serving via WSGI we aren't running this file directly and so our name isn't `__main__`. Otherwise all requests 404 as no endpoints have been defined!

The only part that should be in the main block is the part that starts the Flask dev server, since we don't want this to happen if we're running it via WSGI.

Reply:
As per abae5185f8d88b6c9fc10ac25849628fdd286c7a, I've moved the functions out of that file into a separate file and call them in the top level of `main.py`. This means that tests still work - really so `TEST_BACKEND` from the Flask app config (set only for test clients) can be picked up when running the API for tests. `__name__ == "__main__"` only contains the `app.run()` call so it should work when served via a WSGI server too.